File: Lab12.py
def ejercicio1():
     D = [{'make': 'Nokia', 'model': 216, 'color': 'Black'},
         {'make': 'Apple', 'model': 2, 'color': 'Silver'},
         {'make': 'Huawei', 'model': 50, 'color': 'Gold'},
         {'make': 'Samsung', 'model': 7, 'color': 'Blue'}]
     print(D)  
     continue_ = True
     while continue_:
          key = input("Enter the key to sort by: ")
          if key == 'model':
               D = sorted(D, key=lambda x: x['model'])
          elif key == 'make':
               D = sorted(D, key=lambda x: x['make'])
          elif key == 'color':
               D = sorted(D, key=lambda x: x['color'])
          print(D)
          continue_ = input("Do you want to continue? (y/n): ") == 'y'

def ejercicio2():
     numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
     continue_ = True
     while continue_:
          exponentiation = int(input("Enter the exponentiation: "))
          result = list(map(lambda x: x ** exponentiation, numeros))
          print(result)
          continue_ = input("Do you want to continue? (y/n): ") == 'y'

def ejercicio3():
     x = [[1, 2, 3, 1],
          [4, 5, 6, 0],
          [7, 8, 9, -1]]
     print('Matriz X:\n', x)
     print('Transpuesta de la matriz X usando lambda')
     xT = list(map(lambda x: list(x), zip(*x)))
     print(xT)

def ejercicio4():
     colores = ['rojo', 'verde', 'azul', 'amarillo', 'gris', 'blanco', 'negro']
     colores1 = []
     borrar = True
     print(colores)
     while borrar:
          remove_colores = input('Ingrese un color a borrar:')
          colores1.append(remove_colores)
          print(colores1)
          borrar = input("Do you want to continue? (y/n): ") == 'y'
     print('Lista de colores inicial:', colores)
     print('Lista de colores a eliminar', colores1)
     print('Eliminando colores...')
     # Eliminar elementos de la lista 'colores' usando lambda
     colores_resultantes = list(filter(lambda color: color not in colores1, colores))
     print('Lista de colores resultantes:', colores_resultantes)




def menu():
     continuar = True
     while continuar:
          print('-----Laboratiorio #12-----')
          print('1. Ejercicio 1\n2. Ejercicio 2\n3. Ejercicio 3\n4. Ejercicio 4')
          opcion = input('Ingrese el ejercicio a ver: ')
          if opcion == '1':
               ejercicio1()
          elif opcion == '2':
               ejercicio2()
          elif opcion == '3':
               ejercicio3()
          elif opcion == '4':
               ejercicio4()
          else:
               print('Opcion no valida')
          continuar = input('Desea continuar? (y/n): ') == 'y'

menu()

File: Lab12.hs
import Data.List (sortBy, transpose)
import Data.Ord (comparing)
ejercicio1 :: IO ()
ejercicio1 = do
    let d = [("Nokia", 216, "Black"),
             ("Apple", 2, "Silver"),
             ("Huawei", 50, "Gold"),
             ("Samsung", 7, "Blue")]
    putStrLn "Lista original:"
    print d
    putStrLn "Ingrese la clave para ordenar (model, make, color):"
    key <- getLine
    let sortedD = case key of
                      "model" -> sortBy (comparing (\(_, model, _) -> model)) d
                      "make"  -> sortBy (comparing (\(make, _, _) -> make)) d
                      "color" -> sortBy (comparing (\(_, _, color) -> color)) d
                      _       -> d
    putStrLn "Lista ordenada:"
    print sortedD

ejercicio2 :: IO ()
ejercicio2 = do
    let numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] -- Lista original
    putStrLn "Lista original:"
    print numeros
    putStrLn "Ingrese la potencia a calcular (n):"
    nStr <- getLine                          -- Leer la entrada del usuario como cadena
    let n = read nStr :: Int                 -- Convertir la cadena a un entero
    let result = map (^ n) numeros   -- Aplicar la potencia a cada elemento de la lista
    putStrLn "Resultado:"
    print result                             -- Mostrar el resultado



ejercicio3 :: IO ()
ejercicio3 = do
    let x = [[1, 2, 3, 1],
             [4, 5, 6, 0],
             [7, 8, 9, -1]]
    putStrLn "Matriz original:"
    print x
    putStrLn "Matriz transpuesta:"
    let xT = transpose x
    print xT

ejercicio4 :: IO ()
ejercicio4 = do
    let colores = ["rojo", "verde", "azul", "amarillo", "gris", "blanco", "negro"]
    putStrLn "Lista original:"
    print colores
    putStrLn "Ingrese los colores a borrar separados por comas:"
    input <- getLine
    let colores1 = words [if c == ',' then ' ' else c | c <- input] -- Dividir por comas
    let coloresResultantes = filter (`notElem` colores1) colores   -- Filtrar elementos no deseados
    putStrLn "Lista resultante:"
    print coloresResultantes



menu :: IO ()
menu = do
    putStrLn "----- Laboratorio #12 -----"
    putStrLn "1. Ejercicio 1: Ordenar lista de diccionarios"
    putStrLn "2. Ejercicio 2: Potencia n-ésima de elementos"
    putStrLn "3. Ejercicio 3: Transponer matriz"
    putStrLn "4. Ejercicio 4: Eliminar elementos de una lista"
    putStrLn "Seleccione una opción (1-4):"
    opcion <- getLine
    case opcion of
        "1" -> ejercicio1
        "2" -> ejercicio2
        "3" -> ejercicio3
        "4" -> ejercicio4
        _   -> putStrLn "Opción no válida"
    putStrLn "¿Desea continuar? (y/n):"
    continuar <- getLine
    if continuar == "y" then menu else putStrLn "¡Hasta luego!"

main :: IO ()
main = menu
